File: texpack/sidemap.py
import sys
import Image
import colorsys

class Sidemap:
    def __init__(self):
        pass

    def shift_hue(self, i0, shift):
        width = i0.size[0]
        height = i0.size[1]
        ir = Image.new('RGBA', (width, height))

        for y in xrange(height):
            for x in xrange(width):
                p0 = i0.getpixel((x, y))
                hsl = colorsys.rgb_to_hls(p0[0] / 255.0, p0[1] / 255.0, p0[2] / 255.0)
                huenew = hsl[0] + (shift / 255.0)
                if huenew > 1.0:
                    huenew = huenew - 1.0
                rgb = colorsys.hls_to_rgb(huenew, hsl[1], hsl[2])
                rgba = (int(rgb[0] * 255.0), int(rgb[1] * 255.0), int(rgb[2] * 255.0), p0[3])
                ir.putpixel((x, y), rgba)

        return ir

    def image_add(self, i0, i1):
        width = i0.size[0]
        height = i0.size[1]
        ir = Image.new('RGBA', (width, height))
        for y in xrange(height):
            for x in xrange(width):
                p0 = i0.getpixel((x, y))
                p1 = i1.getpixel((x, y))
                r = p0[0] + p1[0]
                if r > 255:
                    r = 255
                g = p0[1] + p1[1]
                if g > 255:
                    g = 255
                b = p0[2] + p1[2]
                if b > 255:
                    b = 255
                a = p0[3]
                ir.putpixel((x, y), (r, g, b, a))
        return ir

    def image_subtract(self, i0, i1):
        width = i0.size[0]
        height = i0.size[1]
        ir = Image.new('RGBA', (width, height))
        for y in xrange(height):
            for x in xrange(width):
                p0 = i0.getpixel((x, y))
                p1 = i1.getpixel((x, y))
                r = p0[0] - p1[0]
                if r < 0:
                    r = 0
                g = p0[1] - p1[1]
                if g < 0:
                    g = 0
                b = p0[2] - p1[2]
                if b < 0:
                    b = 0
                a = p0[3]
                ir.putpixel((x, y), (r, g, b, a))
        return ir

    def gen_unit_image(self, white, black, hue):
        sidepixels = self.image_subtract(white, black)
        sidecolors = self.shift_hue(sidepixels, hue)
        return self.image_add(black, sidecolors)

    def shift_file(self, whitefile, blackfile, hue, outfile):
        white = Image.open(whitefile).convert('RGBA')
        black = Image.open(blackfile).convert('RGBA')
        result = self.gen_unit_image(white, black, hue)
        result.save(outfile)

File: texpack/SideMap.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;

namespace SpiffCode {

    public static class SideMap {

        public static Bitmap Subtract(Bitmap bm, Bitmap bmBlack) {
            if (bm == null || bmBlack == null)
                return null;
            if (bm.Width != bmBlack.Width || bm.Height != bmBlack.Height)
                return null;

            Bitmap bmr = new Bitmap(bm.Width, bm.Height);

            for (int y = 0; y < bm.Height; y++) {
                for (int x = 0; x < bm.Width; x++) {
                    Color p0 = bm.GetPixel(x, y);
                    Color p1 = bmBlack.GetPixel(x, y);
                    int r, g, b, a;

                    r = p0.R - p1.R;
                    if (r < 0)
                        r = 0;
                    g = p0.G - p1.G;
                    if (g < 0)
                        g = 0;
                    b = p0.B - p1.B;
                    if (b < 0)
                        b = 0;
                    a = p0.A;

                    bmr.SetPixel(x, y, Color.FromArgb(a, r, g, b));
                }
            }

            return bmr;
        }

        public static Bitmap Add(Bitmap bm, Bitmap bmBlack) {
            if (bm == null || bmBlack == null)
                return null;
            if (bm.Width != bmBlack.Width || bm.Height != bmBlack.Height)
                return null;

            Bitmap bmr = new Bitmap(bm.Width, bm.Height);

            for (int y = 0; y < bm.Height; y++) {
                for (int x = 0; x < bm.Width; x++) {
                    Color p0 = bm.GetPixel(x, y);
                    Color p1 = bmBlack.GetPixel(x, y);
                    int r, g, b, a;

                    r = p0.R + p1.R;
                    if (r > 255)
                        r = 255;
                    g = p0.G + p1.G;
                    if (g > 255)
                        g = 255;
                    b = p0.B + p1.B;
                    if (b > 255)
                        b = 255;
                    a = p0.A;

                    bmr.SetPixel(x, y, Color.FromArgb(a, r, g, b));
                }
            }

            return bmr;
        }

        public static Bitmap ShiftHue(Bitmap bm, int shift) {
            if (bm == null)
                return null;

            if (shift == 0)
                return bm;

            Bitmap bmr = new Bitmap(bm.Width, bm.Height);

            for (int y = 0; y < bm.Height; y++) {
                for (int x = 0; x < bm.Width; x++) {
                    Color p0 = bm.GetPixel(x, y);

                    HSL hsl = ColorSys.RGBToHSL(new RGB(p0.R, p0.G, p0.B));
                    hsl.H = hsl.H + shift;
                    RGB rgb = ColorSys.HSLToRGB(hsl);

                    bmr.SetPixel(x, y, Color.FromArgb(p0.A, rgb.R, rgb.G, rgb.B));
                }
            }

            return bmr;
        }

    } // class SideMap

} // namespace SpiffCode
